transcode/transcode.go

```diff
@@ -111,14 +111,57 @@ func RunTranscodeProcess(transcodeRequest TranscodeSegmentRequest, streamName st
 
 	// Generate a unique ID to use when talking to the Broadcaster
 	manifestID := config.RandomTrailer()
-
-	// Iterate through the segment URLs and transcode them
+	// Use channel to queue segments
+	queue := make(chan segmentInfo, len(sourceSegmentURLs))
 	for segmentIndex, u := range sourceSegmentURLs {
-		rc, err := clients.DownloadOSURL(u.URL)
+		queue <- segmentInfo{Input: u, Index: segmentIndex}
+	}
+	close(queue)
+	// Use channel for recieving the errors
+	errors := make(chan error, 100)
+	// Start number of workers in parallel
+	var completed sync.WaitGroup
+	completed.Add(config.TranscodingParallelJobs)
+	for index := 0; index < config.TranscodingParallelJobs; index++ {
+		go transcodeSegment(streamName, manifestID, transcodeRequest, transcodeProfiles, queue, errors, sourceSegmentURLs, targetOSURL, &completed)
```

Comment:
I don't love this pattern of passing channels around, because it makes the control of the flow harder to follow and also makes it difficult to test - I'd prefer to keep `transcodeSegment` as a simple function that knows how to transcode a segment and then have all of the management of the segment queue and allocation of work to goroutines sitting up at this level

Reply:
Also there's enough going on here that it's probably worth moving out of the main Transcode body, so the flow would be something like

<Transcode - high level iterative transcode process> -> <Parallel Segment Transcoder> -> <transcodeSegment>